### DISCUSSION TITLE
Encapsulate traits for reusability

### DISCUSSION THREAD

**Discussion by membla** [GrapesJS/grapesjs#6317] [General]
I have a number of traits that can be applied to a number of components. For some of these traits their values affect other properties of the component, so a kind of computed state. I'd believe this is not a completely uncommon use case.

I'm now thinking about what the best way to encapsulate these traits would be – while the `setValue()` / `getValue()` on `TraitProperties` seem like ideal candidates the fact that they are [not called when the initial value is fetched](https://github.com/GrapesJS/grapesjs/blob/36572c38033022cbf259d4e37fde507f60af4354/packages/core/src/trait_manager/model/Trait.ts#L386) still forces me to use the `init()` on the component to set the initial values of the computed state properties. I then would need to extend the `init()` for each trait I want to add to a component which feels very convoluted.

Am I missing something that would let me fully encapsulate a trait that needs to keep other properties in sync with the current API?

**Comment by artf** [comment]
From what I remember calling custom getValue from the beginning might generate loops in some cases (don't remember which ones though) but I think we could allow this to be controlled with some additional property. Would that solve your use case?

**Reply by membla** [reply]
Right, I think I can see how that could happen. But yes, having `getValue()` always called when accessing the value I believe would allow for a much better encapsulation of traits.
